### PR TITLE
[SymmMem] Allocate symmetric memory through the `CUDACachingAllocator`

### PR DESCRIPTION
**Pull request by Aidyn-A**
This PR makes `symm_mem.empty()` allocate through the `CUDACachingAllocator` when using the implicit symmetric memory pool. This lets symmetric-memory allocations participate in CUDAGraph private-pool reuse, enabling `symm_mem.empty()` inside CUDAGraph capture region.

cc @kwen2501

### CODE REVIEW

File: test/distributed/test_nccl.py [hunk 1]
```diff
@@ -428,6 +427,74 @@ def test_nccl_symmem_collective_cuda_graph(self):
             self.assertEqual(out, torch.full_like(out, expected_sum))
             self.assertEqual(res, out)
 
+    @skip_but_pass_in_sandcastle_if(TEST_WITH_ROCM, "Skip NCCL tests for ROCm")
+    @skip_but_pass_in_sandcastle_if(IS_WINDOWS, "NCCL doesn't support Windows")
+    @requires_nccl_version(
+        (2, 28), "NCCL Symmetric Memory support device API from nccl 2.28"
+    )
+    @skip_if_lt_x_gpu(2)
+    def test_nccl_symmem_tensor_creation_and_collective_cuda_graph(self):
```
Comment (kwen2501) [inline test/distributed/test_nccl.py:436]: nit: would be nice to add a similar test for the CUDA backend :)  Can be in a future PR.

File: torch/distributed/_symmetric_memory/__init__.py [hunk 2]
```diff
@@ -1966,12 +1966,13 @@ def empty(  # type: ignore[misc]
...
-        # Allocate tensor from an implicit memory pool
+        # Allocate through the CUDACachingAllocator so symmetric allocations
+        # can participate in MemPool and CUDAGraph private-pool reuse.
         mempool = get_mem_pool(device)
         # TODO: this path can be made device-agnostic if `use_mem_pool` is
         # elevated from torch.cuda to torch accelerator.
-        with torch.cuda.use_mem_pool(mempool):
-            return _SymmetricMemory.empty_strided_p2p(size, stride, dtype, device)
+        with torch.cuda.use_mem_pool(mempool, device=device):
+            return torch.empty_strided(size, stride, dtype=dtype, device=device)
```
Comment (kwen2501) [inline torch/distributed/_symmetric_memory/__init__.py:1975]: I have been wanting to make that 1-line change too :)

**Review by kwen2501 — APPROVED** [review]
LGTM.
Thanks for making tensor creation + CUDA Graph work!

**Review by kapilsh — APPROVED** [review]
LGTM! thanks for making the change.

File: torch/distributed/_symmetric_memory/__init__.py (hunk 2 above)
Comment (Aidyn-A) [reply torch/distributed/_symmetric_memory/__init__.py:1975]: Hmm, it causes hangs in `distributed/test_symmetric_memory.py`. Debugging...

**Comment by ngimel** [comment]
How is this supposed to work? So we would have nested mempool context managers, symm mem would be allocated in the inner one, it won't be accounted for in graph mem pool, `empty_cache` on the implicit mempool can potentially free the storage? 
Tests don't test this.

File: torch/distributed/_symmetric_memory/__init__.py [hunk 3]
```diff
@@ -1966,12 +1966,13 @@ def empty(  # type: ignore[misc]
...
     if _should_use_implicit_mempool() and device.type == "cuda":
-        # Allocate tensor from an implicit memory pool
+        # Allocate through the CUDACachingAllocator so symmetric allocations
+        # can participate in MemPool and CUDAGraph private-pool reuse.
         mempool = get_mem_pool(device)
         # TODO: this path can be made device-agnostic if `use_mem_pool` is
         # elevated from torch.cuda to torch accelerator.
         with torch.cuda.use_mem_pool(mempool):
-            return _SymmetricMemory.empty_strided_p2p(size, stride, dtype, device)
+            return torch.empty_strided(size, stride, dtype=dtype, device=device)
```
Comment (ngimel) [inline torch/distributed/_symmetric_memory/__init__.py:1975]: so symm_mem.empty never actually went through mempool because mempool is only affecting torch.empty calls, not `_SymmetricMemory.empty_strided_p2p`?

Comment (kwen2501) [inline torch/distributed/_symmetric_memory/__init__.py:1975]: Using `torch.empty_strided` here should be good for memory reuse. With `torch.empty_strided`, the allocation flows through the CUDACachingAllocator, which respects the active pool:

```
symm_mem.empty()
└─ use_mem_pool(mempool)
   └─ torch.empty_strided()
      └─ CUDACachingAllocator::malloc()        ← pool IS consulted now
         │
         ├─ pool has cached block? → return it  (ncclMemAlloc NOT called again)
         │
         └─ pool needs new segment?
            └─ cuda_symm_alloc()
               └─ NCCLSymmetricMemoryAllocator::alloc() → ncclMemAlloc()
                  NCCLAllocation registered in allocations_

del tensor
└─ CUDACachingAllocator::free()
   └─ block → pool free list   (ncclMemFree NOT called)
              NCCLAllocation stays alive in allocations_
```

`NCCLSymmetricMemoryAllocator::free()` is now only triggered when the CUDACachingAllocator evicts a segment from the pool (`cuda_symm_free`), not on every tensor deletion. This is correct for memory reuse: `ncclMemAlloc` is called once per pool segment, and subsequent `symm_mem.empty()` calls of the same size return the cached block without touching NCCL at the allocation level.

**Comment by kwen2501** [comment]
Merging this PR for the MemPool change. Call path verified in the [comment](https://github.com/pytorch/pytorch/pull/187054#discussion_r3449676411) above. 

For full CUDA Graph support, it may be worth more investigation and tests in future PRs. 

**Update**: canceling the merge.

File: test/distributed/test_nccl.py [hunk 4]
```diff
@@ -428,6 +427,69 @@ def test_nccl_symmem_collective_cuda_graph(self):
...
+                )
+                self.assertEqual(out, torch.full_like(out, expected_sum))
+            del inp, out
+        capture_stream.synchronize()
+
+        graph = torch.cuda.CUDAGraph()
+        offset = 13
+        capture_offset.fill_(self.rank + offset)
+        with torch.cuda.graph(graph, stream=capture_stream):
+            inp = symm_mem.empty(numel, dtype=dtype, device=self.device)
```
Comment (wujingyue) [inline test/distributed/test_nccl.py:472]: Thanks for working on this and the change looks good to me! 

One non-blocking question: IIUC, symm_mem.empty allocates from the symmetric memory pool rather than the CUDA graph pool, so it does not provide a stability guarantee. This issue is probably difficult to expose in a unit test because the caching allocator provides stability only on a **best-effort** basis.

Do we have a plan to address this? For example, are we considering a memory pool that both guarantees address stability **and** satisfies the symmetric memory constraints?

File: test/distributed/test_nccl.py [hunk 5]
```diff
@@ -428,6 +427,69 @@ def test_nccl_symmem_collective_cuda_graph(self):
...
+
+        dtype = torch.float
+        numel = 1024
+        capture_offset = torch.tensor(1, dtype=dtype, device=self.device)
+        capture_stream = torch.cuda.Stream(device=self.device)
+
+        # Warm up on the same stream that will be captured. This establishes
+        # NCCL window metadata for the symmetric block that capture reuses.
+        with torch.cuda.stream(capture_stream):
+            inp = symm_mem.empty(numel, dtype=dtype, device=self.device)
```
Comment (wujingyue) [inline test/distributed/test_nccl.py:450]: I couldn't find symm_mem.rendezvous around. Is that not needed?

File: torch/distributed/_symmetric_memory/__init__.py (hunk 3 above)
Comment (Aidyn-A) [reply torch/distributed/_symmetric_memory/__init__.py:1975]: It goes straight to the `empty_strided_p2p`:
https://github.com/pytorch/pytorch/blob/7d067fae44f1495b8c1ff8a3a4868f3782355a49/torch/csrc/distributed/c10d/symm_mem/SymmetricMemory.cpp#L259-L289
That is relying on the `SymmetricMemoryAllocator`:
https://github.com/pytorch/pytorch/blob/7d067fae44f1495b8c1ff8a3a4868f3782355a49/torch/csrc/distributed/c10d/symm_mem/SymmetricMemory.hpp#L99-L119

File: test/distributed/test_nccl.py (hunk 4 above)
Comment (Aidyn-A) [reply test/distributed/test_nccl.py:472]: Yes, I have proposed a PR for `CUDAGraphs` to support multiple memory pools: https://github.com/pytorch/pytorch/pull/187929

File: test/distributed/test_nccl.py (hunk 5 above)
Comment (Aidyn-A) [reply test/distributed/test_nccl.py:450]: Not really, it will be called in `torch.ops.symm_mem.one_shot_all_reduce_out` if needed

File: torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemory.cu [hunk 6]
```diff
@@ -930,9 +938,25 @@ c10::intrusive_ptr<SymmetricMemory> CUDASymmetricMemoryAllocator::rendezvous(
...
+      group->getBackend(c10::DeviceType::CUDA)->getUsePgForSymmMemRendezvous();
+  auto cache_key = group_name_ + (use_pg ? "#pg" : "#store");
+  auto it = block->symm_mems.find(cache_key);
+  uint8_t need_rendezvous = it == block->symm_mems.end();
+  auto peer_needs = storeExchange.all_gather(
+      group->getStore(), group->getRank(), group->getSize(), need_rendezvous);
+  bool should_rendezvous = false;
+  for (auto peer_need : peer_needs) {
+    should_rendezvous |= peer_need != 0;
+  }
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemory.cu:957]: I think symmetric behavior is a user-lib contract. Even in MemPool case. 
We can use the OOB all-gather as a debug mechanism (is something on the list btw), but it is difficult to do it in production runs.

File: torch/distributed/_symmetric_memory/__init__.py (hunk 3 above)
Comment (kwen2501) [reply torch/distributed/_symmetric_memory/__init__.py:1975]: Yeah, I think we should merge this change regardless of CUDA Graphability.

Comment (kwen2501) [reply torch/distributed/_symmetric_memory/__init__.py:1975]: I added tests in this PR to guard the reuse (in both MemPool context case and `symm_mem.empty` case). Do they finish green?

File: torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemory.cu [hunk 7]
```diff
@@ -503,6 +503,14 @@ struct RendezvousRequest {
...
 };
 
+static void reset_signal_pad(const c10::intrusive_ptr<Block>& block) {
+  c10::cuda::CUDAGuard guard(block->device_idx);
+  auto base_addr = reinterpret_cast<uintptr_t>(block->alloc_ref->ptr);
+  auto* signal_pad =
+      reinterpret_cast<void*>(base_addr + block->signal_pad_offset);
+  AT_CUDA_CHECK(cudaMemset(signal_pad, 0, get_signal_pad_size()));
+}
+
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemory.cu:513]: I am 50:50 whether we should do it for the user at the time of allocation. 
Like `torch.empty`, it doesn't guarantee that the returned tensor is filled with 0. 
Would like to defer to @ngimel for advice here.

Comment (ngimel) [reply torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemory.cu:513]: but our existing behavior is that signal pads are filled with 0? I would go with whatever existing behavior is

File: torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemory.cu (hunk 6 above)
Comment (ngimel) [reply torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemory.cu:957]: Agree we should avoid all_gather with StoreExchange, it's very slow (minutes) on a few-K scale jobs

File: torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemory.cu [hunk 8]
```diff
@@ -930,9 +938,25 @@ c10::intrusive_ptr<SymmetricMemory> CUDASymmetricMemoryAllocator::rendezvous(
     group_name_ = *block->default_group_name;
   }
 
-  // If found, this block has been rendezvous by the given group
-  auto it = block->symm_mems.find(group_name_);
-  if (it == block->symm_mems.end()) {
+  reset_signal_pad(block);
```
Comment (ngimel) [inline torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemory.cu:941]: I'm sorry what the hell is this?
It's mildly annoying that we do O(N) search of symmetric memory wiithout a shortcut above, in `find_block_covering`, it was supposed to be a temporary state @kwen2501 
But now EVERY, ABSOLUTELY EVERY rendezvous is turned into a reset memory, exchange handles via store 10-minute exercise? rendezvous is called on every op! It's supposed to have a fast hot path!
Besides, I'm dead against reusing the same symmetric memory for different PGs - if there ever was a recipe for signal pad deadlocking, this is it! We are not serializing symm mem ops on the user promise of good behavior, typically a particular symm mem op goes on the same stream anyway, but this is a giant footgun

**Review by ngimel — CHANGES_REQUESTED** [review]

File: torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemory.cu (hunk 8 above)
Comment (ngimel) [reply torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemory.cu:941]: And I understand that nvidia wants nccl backend to be default for symmetric memory, but it doesn't change the fact that currently the most used backend is CUDA, definitely within meta.

File: torch/distributed/_symmetric_memory/__init__.py (hunk 3 above)
Comment (kapilsh) [inline torch/distributed/_symmetric_memory/__init__.py:1975]: I wrote a quick script to check warmup and capture addresses and corresponding pool ids: https://gist.github.com/kapilsh/f40c5ad21bc70643b86a17c99b131ff2

this is what i get for output which suggest that graph capture pools are empty - this seems weird - just wanted to flag. 

```
NCCL version 2.30.7+cuda13.3
symm pool id          = (0, 1)
graph capture pool id = (1, 0)
warmup symm alloc     -> addr 0xfffa74a00000  pool (0, 1)
in-capture symm alloc -> addr 0xfffa74a00000  pool (0, 1)
symm pool segments          = 1  ['0xfffa74a00000']
graph capture pool segments = 0  []
=> the in-capture symm alloc lives in the symm pool; the graph's own
   capture pool is EMPTY
```

Comment (Aidyn-A) [reply torch/distributed/_symmetric_memory/__init__.py:1975]: I just landed a PR that retains the nested pool for a CUDAGraph:  https://github.com/pytorch/pytorch/pull/187929



If I slightly modify your code to incorporate the changes in #187929 it will output:
```
symm pool id          = (0, 1)
graph capture pool ids = ((1, 0), (0, 1))
warmup symm alloc     -> addr 0xfffb01400000  pool (0, 1)
in-capture symm alloc -> addr 0xfffb01400000  pool (0, 1)
symm pool segments          = 1  ['0xfffb01400000']
graph capture retained pool segments = 1  ['0xfffb01400000']
```

<details><summary>Python script here</summary>
<p>


```python
from __future__ import annotations

import os

import torch
import torch.distributed as dist
import torch.distributed._symmetric_memory as symm_mem

RANK = int(os.environ["RANK"])
LOCAL_RANK = int(os.environ["LOCAL_RANK"])
N = 256


def log0(msg: str) -> None:
    if RANK == 0:
        print(msg, flush=True)


def pool_of(ptr: int) -> tuple[int, int] | None:
    for s in torch.cuda.memory._snapshot()["segments"]:
        if s["address"] <= ptr < s["address"] + s["total_size"]:
            return tuple(s["segment_pool_id"])
    return None


def seg_addrs(pool_id) -> list[int]:
    pid = tuple(pool_id)
    return [
        s["address"]
        for s in torch.cuda.memory._snapshot()["segments"]
        if tuple(s["segment_pool_id"]) == pid
    ]


def main() -> None:
    device = torch.device(f"cuda:{LOCAL_RANK}")
    torch.cuda.set_device(device)
    dist.init_process_group("nccl", device_id=device)
    symm_mem.set_backend("NCCL")
    pool = symm_mem.get_mem_pool(device)

    cs = torch.cuda.Stream(device=device)
    with torch.cuda.stream(cs), torch.cuda.use_mem_pool(pool):
        warm = torch.empty_strided((N,), (1,), dtype=torch.float, device=device)
    warm_addr = warm.data_ptr()
    del warm
    cs.synchronize()
    dist.barrier()

    g = torch.cuda.CUDAGraph()
    with torch.cuda.graph(g, stream=cs):
        with torch.cuda.use_mem_pool(pool):
            sym = torch.empty_strided((N,), (1,), dtype=torch.float, device=device)
        sym.fill_(0.0)

    g_pools = g.pools()

    symm_addrs = seg_addrs(pool.id)
    cap_addrs = seg_addrs(g_pools[1])
    log0(f"symm pool id          = {tuple(pool.id)}")
    log0(f"graph capture pool ids = {tuple(g_pools)}")
    log0(f"warmup symm alloc     -> addr {warm_addr:#x}  pool {tuple(pool.id)}")
    log0(
        f"in-capture symm alloc -> addr {sym.data_ptr():#x}  pool {pool_of(sym.data_ptr())}"
    )
    log0(
        f"symm pool segments          = {len(symm_addrs)}  {[hex(a) for a in symm_addrs]}"
    )
    log0(
        f"graph capture retained pool segments = {len(cap_addrs)}  {[hex(a) for a in cap_addrs]}"
    )

    del g, sym
    torch.cuda.synchronize()
    torch.cuda.empty_cache()
    dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

</p>
</details>

**Comment by Aidyn-A** [comment]
closing in favor of #188650
